### PR TITLE
feat(web): New onboarding

### PR DESCRIPTION
**Pull request by SokratisVidros**
### What changed? Why was the change needed?

Flow
====

1. Run `npx novu@latest dev`
2. Visit http://localhost:2022/studio
3. Redirect to https://web.novu.co/local-studio/auth
4. If signed in, redirect back to http://localhost:2022/studio?local_studio_url=... and load the studio iframe with src = local_studio_url
5. If not signed in, redirect to sign-in or sign-up (consult the nv_onboarding_step cookie) with redirect_url=https://web.novu.co/local-studio/auth
6. When authentication is complete, the browser will redirect to step 3 and replay the following steps.

### CODE REVIEW

File: apps/web/src/components/layout/components/EnsureOnboardingComplete.tsx [hunk 1]
```diff
@@ -1,16 +1,25 @@
...
   const { currentOrganization, environmentId } = useAuth();
 
   if ((!currentOrganization || !environmentId) && location.pathname !== ROUTES.AUTH_APPLICATION) {
     return <Navigate to={ROUTES.AUTH_APPLICATION} replace />;
   }
 
+  const redirectURL = getRedirectURL();
+  if (redirectURL) {
+    // Note: Do not use react-router-dom. The version we have doesn't do instant cross origin redirects.
+    window.location.href = redirectURL;
```
Comment (antonjoel82) [inline apps/web/src/components/layout/components/EnsureOnboardingComplete.tsx]:
```suggestion
    window.location.replace(redirectURL);
```

✏ **suggestion:** Assuming that we don't want the intermediate history if this is an auto-redirect‏

File: apps/web/src/hooks/useAuth.ts [hunk 2]
```diff
@@ -107,9 +107,27 @@ export function useAuth() {
...
 
+  const redirectTo = useCallback(({ url, redirectURL }: { url: string; redirectURL?: string }) => {
+    const finalURL = new URL(url, window.location.origin);
+
+    if (redirectURL) {
+      finalURL.searchParams.append('redirect_url', redirectURL);
+    }
+
+    // Note: Do not use react-router-dom. The version we have doesn't do instant cross origin redirects.
+    window.location.href = finalURL.href;
```
Comment (antonjoel82) [inline apps/web/src/hooks/useAuth.ts]: ❓ **question:** Again here -- do we want to skip the history here by using `replace`?‏

File: apps/web/src/hooks/useAuth.ts [hunk 3]
```diff
@@ -107,9 +107,27 @@ export function useAuth() {
...
+
+    return;
```
Comment (antonjoel82) [inline apps/web/src/hooks/useAuth.ts]:
```suggestion
    return finalURL;
```
or remove the return

File: apps/web/src/hooks/useRedirectURL.ts [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+export function useRedirectURL() {
+  const setRedirectURL = useCallback(() => {
+    const redirectURLFromParams = new URL(window.location.href).searchParams.get(REDIRECT_URL_SEARCH_PARAM) || '';
+
+    // If there is a redirect URL in the URL, set it in the cookie.
+    if (redirectURLFromParams) {
+      // Protect against XSS attacks via the javascript: pseudo protocol.
+      assertProtocol(redirectURLFromParams);
+      // Expires in 7 days.
+      novuRedirectURLCookie.set(redirectURLFromParams, { expires: 7 });
```
Comment (antonjoel82) [inline apps/web/src/hooks/useRedirectURL.ts]: ☑ **chore:** Extract to constant (no magic numbers)‏

File: apps/web/src/pages/studio-onboarding/index.tsx [hunk 5]
```diff
@@ -45,8 +50,8 @@ export const StudioOnboarding = () => {
               marginTop: '50',
             })}
           >
-            To start sending your first workflows, you first need to connect Novu to your Bridge Endpoint. This setup
-            will create a sample Next.js project and pre-configured the @novu/framework client for you.
+            Send your first email notification, by connecting to your Novu Bridge Endpoint. This setup will create a
+            sample Next.js project and with a pre-configured @novu/framework.
```
Comment (antonjoel82) [inline apps/web/src/pages/studio-onboarding/index.tsx]:
```suggestion
            Send your first email notification by connecting to your Novu Bridge Endpoint. This setup will create a
            sample Next.js project with a pre-configured <code>@novu/framework</code>.
```

**Review by scopsy — APPROVED** [review]

File: apps/web/src/pages/studio-onboarding/success.tsx [hunk 6]
```diff
@@ -65,6 +66,10 @@ export const StudioOnboardingSuccess = () => {
 
   useEffect(() => {
     segment.track('Test workflow step completed - [Onboarding - Signup]');
+
+    // Never expires! Well it does, in 10 years but you will change device or browser by then :)
+    novuOnboardedCookie.set('1', { expires: 10 * 365 });
```
Comment (antonjoel82) [inline apps/web/src/pages/studio-onboarding/success.tsx]: ⚠ **issue:** Prefer extracting to a constant.

However, saw this [online](https://stackoverflow.com/a/3290474): 

> As in 2023, obeying the max depends on the web browsers. As of Chrome release M104 (August 2022) cookies can no longer set an expiration date of more than 400 days in the future.‏

File: apps/web/src/studio/LocalStudioAuthenticator.tsx [hunk 7]
```diff
@@ -60,6 +46,34 @@ export function LocalStudioAuthenticator() {
...
+         * After authentication, redirect back to the this /local-studio/auth path and
+         * remember that studio needs to be in onboarding mode.
+         */
+        // currentURL.searchParams.append('studio_path_hint', ROUTES.STUDIO_ONBOARDING);
+
+        return redirectToSignUp({ redirectURL: currentURL.href });
+      }
+
+      return;
+    }
```
Comment (antonjoel82) [inline apps/web/src/studio/LocalStudioAuthenticator.tsx:76]: ☑ **chore:** This `useEffect` is getting pretty huge and hard to follow. Please put a TODO to refactor this to a standalone hook and we can determine how to break it down a bit‏

File: apps/web/src/studio/StudioPageLayout.tsx [hunk 8]
```diff
@@ -1,9 +1,16 @@
...
 import { PrivatePageLayout } from '../components/layout/components/PrivatePageLayout';
 import { LocalStudioPageLayout } from '../components/layout/components/LocalStudioPageLayout';
+import { ROUTES } from '../constants/routes';
+import { Outlet, useLocation } from 'react-router-dom';
 
 export function StudioPageLayout() {
   const state = useStudioState();
+  const { pathname } = useLocation();
+
+  if (pathname.startsWith(ROUTES.STUDIO_ONBOARDING)) {
```
Comment (antonjoel82) [inline apps/web/src/studio/StudioPageLayout.tsx:11]: ☑ **chore:** Extract to helper with other `isStudioRoute` function to encourage people to prevent diverging implementations

File: apps/web/src/utils/string.ts [hunk 9]
```diff
@@ -0,0 +1,5 @@
+export function capitalizeFirstLetter(string) {
+  if (!string) return '';
+
+  return string.charAt(0).toUpperCase() + string.slice(1);
+}
```
Comment (antonjoel82) [inline apps/web/src/utils/string.ts:5]:
```suggestion
export function capitalizeFirstLetter(word: string) {
   return !word ? '' : string.charAt(0).toUpperCase() + string.slice(1);
}
```

💭 **thought (non-blocking):** We should ideally be using CSS for this when possible‏

File: apps/web/src/utils/url.ts [hunk 10]
```diff
@@ -9,3 +9,19 @@ export function validateURL(url: string, msg = 'The provided URL is invalid'): a
     throw new Error(msg);
   }
 }
+
+const ALLOWED_PROTOCOLS = ['http:', 'https:'];
```
Comment (antonjoel82) [inline apps/web/src/utils/url.ts:13]:
```suggestion
const ALLOWED_PROTOCOLS = ['http:', 'https:'] as const;
```

**Comment by scopsy** [comment]
@antonjoel82 thank you for the feedback! Applied most of it, the rest was making some troubles so for the sake of time I will leave it as is for now. We just have a few dependant changes with this PR I would love to resolve 🙏
